src/hazelcore/semantics/TextShape.re

```diff
@@ -9,34 +9,46 @@ type t =
   | ExpandingKeyword(ExpandingKeyword.t)
   | Var(Var.t);
 
+let int_of_intlit_opt = (n: string): option(int) => {
+  let stripped_text = StringUtil.strip_underscore(n);
+  int_of_string_opt(stripped_text);
+};
+
+let float_of_floatlit_opt = (f: string): option(float) => {
+  let stripped_text = StringUtil.strip_underscore(f);
+  float_of_string_opt(stripped_text);
+};
+
 let of_text = (text: string): option(t) => {
+  let stripped_text = StringUtil.strip_underscore(text);
   switch (
-    int_of_string_opt(text),
-    float_of_string_opt(text),
+    int_of_string_opt(stripped_text),
+    float_of_string_opt(stripped_text),
     bool_of_string_opt(text),
     ExpandingKeyword.mk(text),
   ) {
   /* the num_digits double count with leading zeros when the int is evaluated to 0  */
   | (Some(n), _, _, _) when n == 0 =>
```

Comment:
you don't need "when n == 0", you can just match on "Some(0)", though I'm not entirely sure why 0 is special cased like this?